pkg/sql/schemachanger/scexec/scmutationexec/constraint.go

```diff
@@ -169,20 +169,125 @@ func (m *visitor) RemoveForeignKeyConstraint(
 	if err != nil || out.Dropped() {
 		return err
 	}
+	var found bool
 	for i, fk := range out.OutboundFKs {
 		if fk.ConstraintID == op.ConstraintID {
 			out.OutboundFKs = append(out.OutboundFKs[:i], out.OutboundFKs[i+1:]...)
-			return nil
+			found = true
+			break
 		}
 	}
 	for i, m := range out.Mutations {
 		if c := m.GetConstraint(); c != nil &&
-			c.ConstraintType != descpb.ConstraintToUpdate_FOREIGN_KEY &&
+			c.ConstraintType == descpb.ConstraintToUpdate_FOREIGN_KEY &&
 			c.ForeignKey.ConstraintID == op.ConstraintID {
 			out.Mutations = append(out.Mutations[:i], out.Mutations[i+1:]...)
-			return nil
+			found = true
+			break
+		}
+	}
+	if !found {
+		return errors.AssertionFailedf("foreign key with ID %d not found in origin table %q (%d)",
+			op.ConstraintID, out.GetName(), out.GetID())
+	}
+	return nil
```

Comment:
nit: fyi you could probably do away with `found` and `break` etc. with a `defer`, same applies to the ops below; not sure if it's worth it though. This is fine btw.